patches/api/0328-Add-Raw-Byte-Entity-Serialization.patch

```diff
@@ -3,28 +3,130 @@ From: Mariell Hoversholm <proximyst@proximyst.com>
  
-+    byte[] serializeEntity(org.bukkit.entity.Entity entity);
++    /**
++     * Serializes the provided entity.
++     *
++     * @param entity entity
++     * @return serialized entity data
++     * @see #serializeEntity(org.bukkit.entity.Entity, io.papermc.paper.entity.EntitySerializationFlag...)
++     * @see #deserializeEntity(byte[], World, boolean, boolean)
++     */
++    default byte[] serializeEntity(@org.jetbrains.annotations.NotNull org.bukkit.entity.Entity entity) {
++        return serializeEntity(entity, new io.papermc.paper.entity.EntitySerializationFlag[0]);
++    }
 +
-+    default org.bukkit.entity.Entity deserializeEntity(byte[] data, World world) {
++    /**
++     * Serializes the provided entity.
++     *
++     * @param entity entity
++     * @param serializationFlags serialization flags
++     * @return serialized entity data
++     * @see #deserializeEntity(byte[], World, boolean, boolean)
++     */
++    byte[] serializeEntity(@org.jetbrains.annotations.NotNull org.bukkit.entity.Entity entity, @org.jetbrains.annotations.NotNull io.papermc.paper.entity.EntitySerializationFlag... serializationFlags);
++
++    /**
++     * Deserializes the entity from data.
++     * The entity's {@link java.util.UUID} as well as passengers will not be preserved.
++     *
++     * @param data serialized entity data
++     * @param world world
++     * @return deserialized entity
++     *
++     * @see #deserializeEntity(byte[], World, boolean, boolean)
++     * @see #serializeEntity(org.bukkit.entity.Entity, io.papermc.paper.entity.EntitySerializationFlag...)
++     * @see org.bukkit.entity.Entity#spawnAt(Location, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason)
++     */
++    default org.bukkit.entity.Entity deserializeEntity(byte[] data, @org.jetbrains.annotations.NotNull World world) {
 +        return deserializeEntity(data, world, false);
 +    }
 +
-+    org.bukkit.entity.Entity deserializeEntity(byte[] data, World world, boolean preserveUUID);
++    /**
++     * Deserializes the entity from data.
++     * The entity's passengers will not be preserved.
++     *
++     * @param data serialized entity data
++     * @param world world
++     * @param preserveUUID whether to preserve uuid
++     * @return deserialized entity
++     *
++     * @see #deserializeEntity(byte[], World, boolean, boolean)
++     * @see #serializeEntity(org.bukkit.entity.Entity, io.papermc.paper.entity.EntitySerializationFlag...)
++     * @see org.bukkit.entity.Entity#spawnAt(Location, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason)
++     */
++    default org.bukkit.entity.Entity deserializeEntity(byte[] data, World world, boolean preserveUUID) {
++        return deserializeEntity(data, world, preserveUUID, false);
++    }
++
++    /**
++     * Deserializes the entity from data.
++     *
++     * @param data serialized entity data
++     * @param world world
++     * @param preserveUUID whether to preserve uuid
```

Comment:
I wonder if preserveUUID in this case should also apply to passangers.
In my opinion it very much should, as this method may be used to clone entities via the API and removing and regenerating the uuid of entities is required for said behaviour.

Reply:
It currently does apply to both the entity and its passengers. I guess this could be clarified in the javadoc?
One thing I still wonder about is whether non-persistent passengers should be saved with the `FORCE` flag?

Reply:
Yea sorry, didn't see the helper method.
Javadocs should certainly reflect it.
IMO, if possible, FORCE should probably allow passangers that are non-persistent to be stored.